Fix default set class attribute to None

arwadevel opened the pull request:
**Problem:**
Dictionaries ({}) are mutable. If an instance modifies formset_kwargs, the change is visible to all instances of the class because they share the same object.
**Solution:**
None is immutable. If an instance modifies form_class, a new instance-specific attribute is created, leaving other instances unaffected.
**Corrected Implementation:**
```
class BaseFormSetFactory:
    formset_kwargs = None  # Immutable default value

    def __init__(self):
        self.formset_kwargs = {}  # Each instance gets its own dictionary
        
```
**Key Points:**

- Mutable default values (like {} or []) in class attributes can lead to unintended shared state.

- Immutable defaults (like None) are safe and avoid shared modifications.

- Mutable attributes should be initialized in __init__ to ensure independence across instances.

flemmarwa committed:
Fix default set class attribute to None

sdolemelipone commented:
Hi @arwadevel, thanks for taking the time to contribute. I think there are a couple of reasons why this change isn't right for the library:

Regarding the dictionaries, the intended use of this library is that in most cases, for convenience, keyword arguments should be defined at the class level, i.e.
```python
class MyFormSetFactory(BaseFormSetFactory):
    ...
    factory_kwargs = {"extra": 3}
```
The brevity of this definition is part of the library's intended utility. Overriding `.get_factory_kwargs()` is only required if arguments need to be adjusted at run time. A mutable class level attribute is a necessary evil in exchange for ease of use.

The class level definitions serve as a useful hint as to what type of variable should be used for each class level attribute.

It's a shame that python doesn't have type hinting for class level attributes. It's also a shame that it doesn't have a builtin, neat notation for a frozen dictionary.

Regarding the lists, there is a similar point to be made. There's an argument that they could be replaced with tuples `()`. However I don't think it's uncommon to define a list at class level, e.g.
```python
class MyNamedModelFormSet(NamedFormsetMixin, CreateWithInlinesView):
    ...
    inlines = MyFormSet
    inlines_names = ["my_formset"]
```
and then call `.append()` on the list later on when dynamically adding values.

So, we do the best we can which is to call `[:]` and `.copy()` where necessary, and leave it to the user to do the same in any subclass code if they access the class level attributes directly. This isn't ideal but it is consistent.

Happy to hear any counterpoints to the above.

arwadevel commented:
Thanks for the quick reply @sdolemelipone 

I understand your position but I would like to better explain the problem that arose in my use.
While developing with` django-extra-views`, I realized that the initial approach to override the `queryset` of an `InlineFormSet` element in the `__init__` method introduced a concurrency issue. 
Initially, I implemented the logic as follows:
```
class MyInlineFormSetFactory(InlineFormSetFactory):
    def __init__(self, parent_model, request, instance, view_kwargs=None, view=None, initial=None):
        super().__init__(parent_model, request, instance, view_kwargs, view)
        if instance is not None and instance.pk is not None:
            if hasattr(self, "get_queryset") and callable(self.get_queryset):
                self.formset_kwargs["queryset"] = self.get_queryset()
```
**Problem Identified:**
Initially I hadn't seen the get_formset_kwargs and it seemed more intuitive to force the `__init__`.
The issue was that `formset_kwargs` is a mutable class attribute. Modifying it in `__init__` caused concurrency problems among instances, as they all shared the same object. This meant that changes made by one instance affected others, leading to unexpected behavior.

**My current solution:**
I fixed the issue by moving the queryset override logic to the get_formset_kwargs method, which is called at runtime for each instance. This way, each instance manages its own queryset without interfering with others:
```
class MyInlineFormSetFactory(InlineFormSetFactory):
    def get_formset_kwargs(self):
        if self.object is not None and self.object.pk is not None:
            if hasattr(self, "get_queryset") and callable(self.get_queryset):
                self.formset_kwargs["queryset"] = self.get_queryset()
        return super().get_formset_kwargs()
```
**Motivation for the Proposal:**
My pull request proposes avoiding mutability at the class level for formset_kwargs. Keeping the class attribute definitions as None would eliminate concurrency issues in the Mixin class.
This approach:
- Eliminates concurrency issues between instances.
- Maintains flexibility for users who want to customize attributes defined at top of Mixin class.
- Respects the principle of encapsulation, as each instance manages its own state independently.

sdolemelipone commented:
The recommended way to override `formset_kwargs` at run time is:

```python
class MyInlineFormSetFactory(InlineFormSetFactory):
    def get_formset_kwargs(self):
        formset_kwargs = super().get_formset_kwargs()
        if self.object is not None and self.object.pk is not None:
            if hasattr(self, "get_queryset") and callable(self.get_queryset):
                formset_kwargs["queryset"] = self.get_queryset()
        return formset_kwargs
```
The call to `super().get_formset_kwargs` ensures that `self.formset_kwargs.copy()` is returned, which means any further modifications do not affect the class level attribute.

That should address any issues with modification at run time.
